sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala

```diff
@@ -308,12 +311,97 @@ object PreProjectSplitIterator {
+            }
           }.getOrElse {
             GpuBatchUtils.minGpuMemory(dataType, true, numRows)
           }
         }
     }.sum
   }
+
+  private def calcMemorySizeForLiteral(litVal: Any, litType: DataType, numRows: Int): Long = {
+    // One literal size = value size + additional buffers size (offset and/or validity)
+    val litSize = calcLitValueSize(litVal, litType) + {
+      val pickRowNum: Int => Int = rowNum => if (litVal == null) 0 else rowNum
+      litType match {
+        case ArrayType(elemType, hasNullElem) =>
+          val numElems = pickRowNum(litVal.asInstanceOf[ArrayData].numElements())
+          // A GPU array literal requires only one column as the child
+          estimateLitAdditionSize(hasNullElem, hasOffset(elemType), numElems)
+        case StructType(fields) =>
+          val childrenNumRows = pickRowNum(1)
+          // A GPU struct literal requires "fields.size" columns as the children.
+          fields.map(f =>
+            estimateLitAdditionSize(f.nullable, hasOffset(f.dataType), childrenNumRows)
+          ).sum
+        case MapType(keyType, valType, hasNullValue) =>
+          val mapRowsNum = pickRowNum(litVal.asInstanceOf[MapData].numElements())
+          // A GPU map literal requires 4 columns as the children.
+          // the key and value column, along with two wrapper columns as below
+          //   " list<struct{key, value}> "
+          estimateLitAdditionSize(false, hasOffset(keyType), mapRowsNum) + // key
+            estimateLitAdditionSize(hasNullValue, hasOffset(valType), mapRowsNum) + // value
+              estimateLitAdditionSize(false, false, mapRowsNum) + // struct
+                estimateLitAdditionSize(false, true, mapRowsNum) // top list
+        case _ => 0L // primitive types has no nested additional buffers
+      }
+    }
+    // totalSize = litSize * numRows + top additional buffers size after expanding to a column.
+    litSize * numRows + estimateLitAdditionSize(litVal == null, hasOffset(litType), numRows)
+  }
+
+  private def estimateLitAdditionSize(hasNull: Boolean, hasOffset: Boolean, rows: Int): Long = {
+    // Additional buffers size, it is not nested for literals,
+    // so no need to do it recursively.
+    var totalSize = 0L
+    if (hasNull) {
+      totalSize += GpuBatchUtils.calculateValidityBufferSize(rows)
+    }
+    if (hasOffset) {
+      totalSize += GpuBatchUtils.calculateOffsetBufferSize(rows)
+    }
+    totalSize
+  }
+
+  private def calcLitValueSize(lit: Any, litTp: DataType): Long = if (lit == null) {
+    if (GpuBatchUtils.isFixedWidth(litTp)) {
+      litTp.defaultSize
+    } else {
+      0L
+    }
+  } else {
+    litTp match {
+      case StringType => lit.asInstanceOf[UTF8String].numBytes()
+      case BinaryType => lit.asInstanceOf[Array[Byte]].length
+      case ArrayType(elemType, _) =>
+        lit.asInstanceOf[ArrayData].array.map(calcLitValueSize(_, elemType)).sum
```

Comment:
This appears to only work on UnsafeArrayData. If I test with the following it fails. (I opened up the API to be public so I could test things manually)

```
scala> PreProjectSplitIterator.calcMemorySizeForLiteral(ArrayData.toArrayData(Array(1L)), ArrayType(LongType), 100)
java.lang.UnsupportedOperationException: Not supported on UnsafeArrayData.
  at org.apache.spark.sql.catalyst.expressions.UnsafeArrayData.array(UnsafeArrayData.java:103)
  at com.nvidia.spark.rapids.PreProjectSplitIterator$.calcLitValueSize(basicPhysicalOperators.scala:381)
  at com.nvidia.spark.rapids.PreProjectSplitIterator$.calcMemorySizeForLiteral(basicPhysicalOperators.scala:328)
  ... 51 elided
```